File: frontend/src/api/projects.js
import apiClient from "./client";

export const createProject = async (projectData) => {
  const response = await apiClient.post("/api/projects/", projectData);
  return response.data;
};

export const listProjects = async () => {
  const response = await apiClient.get("/api/projects/");
  return response.data;
};

export const getProject = async (projectId) => {
  const response = await apiClient.get(`/api/projects/${projectId}`);
  return response.data;
};

export const updateProject = async (projectId, projectData) => {
  const response = await apiClient.patch(`/api/projects/${projectId}`, projectData);
  return response.data;
};

export const deleteProject = async (projectId) => {
  const response = await apiClient.delete(`/api/projects/${projectId}`);
  return response.data;
};

File: backend/app/api/projects.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List

from app.core.database import get_db
from app.schemas.project_schema import ProjectRead, ProjectCreate, ProjectUpdate
from app.services import project_service

router = APIRouter()

@router.post("/", response_model=ProjectRead)
def create_project(project_in: ProjectCreate, db: Session = Depends(get_db)):
    return project_service.create_project(db, project_in)

@router.get("/", response_model=List[ProjectRead])
def list_projects(skip: int = 0, limit: int = 100, db: Session = Depends(get_db)):
    return project_service.list_projects(db, skip=skip, limit=limit)

@router.get("/{project_id}", response_model=ProjectRead)
def get_project(project_id: int, db: Session = Depends(get_db)):
    project = project_service.get_project(db, project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    return project

@router.patch("/{project_id}", response_model=ProjectRead)
def update_project(project_id: int, project_in: ProjectUpdate, db: Session = Depends(get_db)):
    project = project_service.get_project(db, project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    return project_service.update_project(db, project, project_in)

@router.delete("/{project_id}", status_code=204)
def delete_project(project_id: int, db: Session = Depends(get_db)):
    success = project_service.delete_project(db, project_id)
    if not success:
        raise HTTPException(status_code=404, detail="Project not found")
    return None
